File: CSharpCodeBase/unity/unityentity.lua
local Class = require 'utils.hump.class'
local Vector2 = require 'utils.hump.vector2'
local Vector3 = require 'utils.hump.vector3'
local Entity = require 'unity.entity'

local UnityExistsEntity = Class({__includes=Entity, init = function (self, gameObject)      
      Entity.init(self)      
      self.gameObject = gameObject
      self.name = gameObject.name
      self:InitRadiusAndHeight(gameObject)      
      self.transform = self.gameObject.transform
      self.luaMapper = self.gameObject:GetComponent("LuaMapper")
      self.luaMapper:SetEntity(self)
      
      -- Public api fields
      self.static = true
      self.visible = false
      self.enabled = false
      self.interactable = true
      
      self._cachedPosition = nil
      self._cachedPositionValid = false
      self.visibilityTime = 0
end})

function UnityExistsEntity:InitRadiusAndHeight(gameObject)
  local result = {}
  RigidbodyUtils.GetCapsuleColliderData(gameObject, result)
  self.radius = result.radius or 0.5
  self.height = result.height or 0.5
  self.capsuleOffset = Vector3(result.ox or 0, result.oy or 0, result.oz or 0)
end

function UnityExistsEntity:OnChangeVisibility(state)
  self.visible = state
  self.enabled = state
  self:Message("OnChangeVisibility", state)
  self.visibilityTime = 0
end

function UnityExistsEntity:Update()
  if self.visible then
    self.visibilityTime = self.visibilityTime  + GameController.deltaTime
  end
  if not self.static then self._cachedPositionValid = false end
  Entity.Update(self)
end

function UnityExistsEntity:OnUnload()
  self:Message('RemoveFromTimeService')
  GameController.entityFactory:Destroy(self)
  --GameController:RemoveEntity(self)
  --GameController.pooler:Release(self.gameObject)
end

function UnityExistsEntity:FixedUpdate()
  if not self.static then self._cachedPositionValid = false end
  Entity.FixedUpdate(self)
end

function UnityExistsEntity:LateUpdate()
  if not self.static then self._cachedPositionValid = false end
  Entity.LateUpdate(self)
end

function UnityExistsEntity:OnEvent(data)
  self:Message("OnEvent", data)
end

function UnityExistsEntity:Destroy()
  GameController.entityFactory:Destroy(self)
end

function UnityExistsEntity:GetType()
  return "UnityExistsEntity"
end

function UnityExistsEntity:LookAt(vec3)
  local rot = RotationUtils.LookRotation(Vector3(vec3.x, 0, vec3.z))
  self:SetRotation(rot)
end

function UnityExistsEntity:SetPosition(vec3)
  self._cachedPosition = vec3
  Transform.SetPosition(self.transform, vec3)
end

function UnityExistsEntity:ResetPositionCache()
  self._cachedPositionValid = false
end

function UnityExistsEntity:GetPosition(misscache)
  local nocache = misscache or false
  if not self._cachedPositionValid or nocache then
    local x = self.luaMapper:_X()
    local y = self.luaMapper:_Y()
    local z = self.luaMapper:_Z()
    self._cachedPosition = Vector3(x,y,z)
    self._cachedPositionValid = true
  end
  return self._cachedPosition
end

function UnityExistsEntity:GetForwardVector()
  return Transform.GetForwardVector(self.transform)
end

function UnityExistsEntity:GetPosition2D()
  local pos = self:GetPosition()
  return Vector2(pos.x, pos.z)
end

function UnityExistsEntity:SetRotation(vec3)
  Transform.SetRotation(self.transform, vec3)
end

function UnityExistsEntity:GetRotation()
  local x = self.luaMapper:_RX()
  local y = self.luaMapper:_RY()
  local z = self.luaMapper:_RZ()
  return Vector3(x,y,z)  
end

function UnityExistsEntity:GetEffectiveDistanceToVec(vec3)
  local localPos = self:GetPosition2D()
  local targetPos = Vector2(vec3.x, vec3.z)
  local result = (localPos - targetPos):Length() - self.radius
  Console:Message(result)
  if result < 0 then result = 0 end
  return result
end

function UnityExistsEntity:GetSimpleDistance(targetEntity)
  local localPos = self:GetPosition2D()
  local targetPos = targetEntity:GetPosition2D()
  local result = (localPos - targetPos):Length()
  return result
end

function UnityExistsEntity:GetSimpleDistanceToVec(vec3)
  local localPos = self:GetPosition2D()
  local targetPos = Vector2(vec3.x, vec3.z)
  local result = (localPos - targetPos):Length()
  return result
end

function UnityExistsEntity:GetEffectiveDistance(targetEntity)  
  local localPos = Transform.TransformPoint(self.transform, self.capsuleOffset)
  local targetPos = Transform.TransformPoint(targetEntity.transform, targetEntity.capsuleOffset)
  localPos.y = 0
  targetPos.y = 0
  local result = (localPos - targetPos):Length() - self.radius - targetEntity.radius
  if result < 0 then result = 0 end
  return result 
end

function UnityExistsEntity:SetScale(vec3)
  Transform.SetScale(self.transform, vec3)
end

function UnityExistsEntity:ChangeRadiusAndHeight(gameObject)
   self.radius = RigidbodyUtils.TryGetRadius(gameObject) or 0.5
   self.height = RigidbodyUtils.TryGetHeight(gameObject) or 0.5
end 

return UnityExistsEntity

File: CSharpCodeBase/unity/unityentity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;
using System.Collections;
using MainGame.core;


namespace MainGame {
    public class UnityEntity : MonoBehaviour {
        protected bool visible;
        protected bool enabled;
        protected bool interactable;
        protected bool isStatic;
        protected float visibilityTime;
        protected float radius;
        protected float height;
        protected Vector3 capsuleOffset;

        public bool IsInteractable {
            get { return interactable; }
            set { interactable = value; }
        }

        public void Awake() {
            InitRadiusAndHeight();

            this.isStatic = true;
            this.enabled = false;
            this.visible = false;
            this.interactable = true;
            this.visibilityTime = 0;
        }

        public void FixedUpdate()
        {
        }

        //FIXME PLEASE
        public void InitRadiusAndHeight() {
            //var result = {}
            //RigidbodyUtils.GetCapsuleColliderData(gameObject, result);
            //this.radius = result.radius  ||  0.5;
            //this.height = result.height  ||  0.5;
            //this.capsuleOffset = Vector3(result.ox  ||  0, result.oy  ||  0, result.oz  ||  0);
            this.radius = 0.5f;
            this.height = 0.5f;
            this.capsuleOffset = Vector3.zero;
        }

        public void OnChangeVisibility(bool state) {
            this.visible = state;
            this.enabled = state;
            SendMessage("OnChangeVisibility", state);
            this.visibilityTime = 0;
        }

        public void Update() {
            if (this.visible) {
                this.visibilityTime = this.visibilityTime + Time.deltaTime;
            }
        }

        //FIXME
        public void OnUnload() {
            SendMessage("RemoveFromTimeService");
            //GameController.entityFactory:Destroy(self);
            //GameController:RemoveEntity(self)
            //GameController.pooler:Release(this.gameObject)
        }

        public void OnEvent(object[] data) {
            SendMessage("OnEvent", data);
        }

        //FIXME
        public void Destroy() {
            //GameController.entityFactory:Destroy(self);
        }

        public string GetType() {
            return "UnityEntity";
        }

        //FIXME
        public void LookAt(Vector3 vector) {
            //var rot = RotationUtils.LookRotation(Vector3(vec3.x, 0, vec3.z));
            //self:SetRotation(rot);
        }

        public void SetPosition(Vector3 position) {
            transform.position = position;
        }

        public Vector3 GetPosition() {
            return transform.position;
        }

        public Vector2 GetPosition2D() {
            var pos = this.GetPosition();
            return new Vector2(pos.x, pos.z);
        }

        public void SetRotation(Vector3 rotation) {
            transform.rotation = Quaternion.Euler(rotation);
        }

        public Vector3 GetRotation() {
            return transform.rotation.eulerAngles;
        }


        public float GetEffectiveDistanceToVec(Vector3 vec3) {
            var localPos = GetPosition2D();
            var targetPos = new Vector2(vec3.x, vec3.z);
            var length = (localPos - targetPos).magnitude;
            var result = length - this.radius;
            Debug.Log(result);
            if (result < 0) {
                result = 0;
            }
            return result;
        }

        public float GetSimpleDistance(UnityEntity targetEntity) {
            var localPos = GetPosition2D();
            var targetPos = targetEntity.GetPosition2D();
            var result = (localPos - targetPos).magnitude;
            return result;
        }

        public float GetSimpleDistanceToVec(Vector3 vec3) {
            var localPos = GetPosition2D();
            var targetPos = new Vector2(vec3.x, vec3.z);
            var result = (localPos - targetPos).magnitude;
            return result;
        }

        public float GetEffectiveDistance(UnityEntity targetEntity) {
            var localPos = transform.TransformPoint(this.capsuleOffset);
            var targetPos = transform.TransformPoint(targetEntity.capsuleOffset);
            localPos.y = 0;
            targetPos.y = 0;
            var result = (localPos - targetPos).magnitude - this.radius - targetEntity.radius;
            if (result < 0) { result = 0; }
            return result;
        }

        public void SetScale(Vector3 scale) {
            transform.localScale = scale;
        }

        //FIXME
        public void ChangeRadiusAndHeight(GameObject gameObject) {
            //this.radius = RigidbodyUtils.TryGetRadius(gameObject)  ||  0.5;
            //this.height = RigidbodyUtils.TryGetHeight(gameObject)  ||  0.5;
        }
    }
}